src/native/eventpipe/ds-protocol.c

```diff
@@ -306,6 +314,50 @@ ipc_message_flatten (
 	ep_exit_error_handler ();
 }
 
+static
+bool
+ipc_message_try_parse_string_utf16_t_byte_array (
```

Comment:
@lateralusX Is this mostly meant as a targeted fix? Feels odd we are only safekeeping this one case. I wonder if we can statically test for alignment on these reads or check dynamically if the platform supports it and perform copies when necessary. This also feels like a potential pain point for our portable builds (pretty much would for reads to always be aligned). The fix itself (copying) LGTM, just feels like we might be chasing bugs in this area for a bit.

Reply:
First, fix was evaluating if it was the root cause of the ARM 32-bit issue seen on CoreCLR. Second I didn't want to switch to copy semantics in all cases, since it has been like that for a long time (inherited from C++ codebase) and its an optimization that can be used in most cases of current parsing logic (all except in the collect trace 2). Switching to copy semantics would also change the ownership rules of parsed UTF16 strings meaning we would need to do more adjustments in parsing logic artifacts that currently don't have ownership of the UTF16 strings (see). I added a runtime assert to make sure the version not doing a byte copy, `ds_ipc_message_try_parse_string_utf16_t`, will enforce that the passed in buffer is property aligned, in order to catch any future unaligned use. I also looked through all existing calls to `ds_ipc_message_try_parse_string_utf16_t` and they all currently parsed on aligned buffer address (except the fixed collect trace 2 parsing logic) The above function, `ipc_message_try_parse_string_utf16_t_byte_array` is also made static, so only called from within that source. Doing a copy if necessary logic would even complicate the ownership rules of these strings more.

Since the alignment is based on the parsing logic implemented in functions like `eventpipe_collect_tracing2_command_try_parse_payload` there is no static test, but as I said above, I added a runtime validation on debug builds, so as long as we have tests covering this (that we apparently didn't have until we updated the Diagnostic Client), we should hit assert on debug builds, if we add new payloads that could trigger the same issue in future.

The actual SIGBUS happening in this case is also rather interesting since you could argue that it is a compiler optimization bug. ARM supports unaligned reads for word and halfword instructions, but not for double word. The code does two word size reads inside [UTF](https://github.com/dotnet/runtime/blob/4da6b9a8d55913c0ea560d63590d35dc942425be/src/coreclr/pal/src/locale/utf8.cpp#L2523), that would have worked if optimizer didn't bundle them in to a LDRD instruction that will need address to be aligned, so optimization moved from a program that would have worked to one that triggers a SIGBUS.

So to sum up, current fix take care of the scenario currently hit with unaligned UTF16 string in IPC message payload. I added an assert to prevent future incorrect use of UTF16 string parsing function using unaligned addresses. Alternative would be to abandon idea to reuse UTF16 string from payload and always take copy, totally doable, but will add more changes to all current *CommandPayload structs, since they will need to switch to owning copy of UTF16 string and make sure they are freed as part of their fini/free implementations.